pkg/internal/apis/acme/validation/order.go

```diff
@@ -38,7 +38,13 @@ func ValidateOrderUpdate(_ *admissionv1.AdmissionRequest, oldObj, newObj runtime
 	el := field.ErrorList{}
 	el = append(el, ValidateOrderSpecUpdate(old.Spec, new.Spec, field.NewPath("spec"))...)
 	el = append(el, ValidateOrderStatusUpdate(old.Status, new.Status, field.NewPath("status"))...)
-	return el, nil
+	warnings := validateAPIVersion(a.Kind)
+	return el, warnings
+}
+
+func ValidateOrder(a *admissionv1.AdmissionRequest, obj runtime.Object) (field.ErrorList, validation.WarningList) {
+	warnings := validateAPIVersion(a.Kind)
```

Comment:
I think we should be using `RequestKind` instead of `Kind`: https://github.com/kubernetes/api/blob/532146f2aa2a309105e43a4c7e7b2fb94341aabd/admission/v1beta1/types.go#L59-L70

The `kind` field indicates the GVK of the resource submitted to the webhook, whereas `requestKind` denotes the GVK of the resource submitted to the apiserver originally.

The reason this works still today is in our validating (and mutating) webhook, we denote `apiVersions` as `*` (see https://github.com/jetstack/cert-manager/blob/e30dae687a0b498b91b0bb6df7fa51a700241fdb/deploy/charts/cert-manager/templates/webhook-validating-webhook.yaml#L33-L34) - this is indicating that our webhook is able to decode/understand _all_ API versions of resources. This is _technically_ bad practice, because if we were to introduce a new API version and the webhook has not been updated beforehand, the webhook would fail to process the request.

Instead, it'd be typical to denote `apiVersions` as just `v1`, and then have the webhook *expect* a v1 resource to be submitted to it during validation/mutation. That said, it's not _too much_ of an issue in our case because conversion is done by the same webhook anyway, so it sort of just avoids the extra hop/round trip to the webhook for conversion ahead of admission.

But, for correctness, we should be using `RequestKind` when printing deprecation warnings or any other kind of validation :)

Reply:
Thanks for that description! I was trying to understand when the conversion happens whilst working on this PR and couldn't fully figure it out.
I guess we could change the apiVersions for validating webhook in another PR?
I've changed `Kind` -> `RequestKind`

Reply:
> I guess we could change the apiVersions for validating webhook in another PR?

That makes sense :) I think switching the `apiVersions` first (possibly for v1.4?) would be a good idea, and then in v1.5 we can remove the logic inside of the webhook that handles conversion during the *admission* handlers. The idea of having a two-phase approach would be to avoid race issues if the webhook binary is updated to only support v1 *before* the Validating/Mutating webhook configuration resources are updated to not include `*`

Reply:
>  in v1.5 we can remove the logic inside of the webhook that handles conversion during the admission handlers

What would be the reason for removing the conversion before stopping to serve the APIs (in `v1.6`)?